test/core/security/connection_auth_context_test.cc

```diff
@@ -0,0 +1,105 @@
+//
+//
+// Copyright 2024 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "src/core/lib/security/context/connection_auth_context.h"
+
+#include <gtest/gtest.h>
+
+#include <cstdint>
+
+#include "src/core/util/orphanable.h"
+#include "test/core/test_util/test_config.h"
+
+namespace grpc_core {
+namespace {
+
+class Foo {
+ public:
+  explicit Foo(const char* value) : value_(value) {}
+  const char* value() const { return value_; }
+
+ private:
+  const char* value_;
+};
+
+struct FooAuthProperty {
+  using ValueType = Foo;
+  static const char* name() { return name_; }
+  static uint16_t id() { return id_; }
+  static void Destroy(Foo* value) { delete value; }
+  static Foo* Construct(const char* value) { return new Foo(value); }
+
+ private:
+  static uint16_t id_;
+  static constexpr const char* name_ = "grpc.foo.auth.property";
+};
+
+class Bar {
+ public:
+  explicit Bar(int value) : value_(value) {}
+  int value() const { return value_; }
+
+ private:
+  int value_;
+};
+
+struct BarAuthProperty {
+  using ValueType = Bar;
+  static const char* name() { return name_; }
+  static uint16_t id() { return id_; }
```

Comment:
for arena contexts I added `ArenaContextTraits` to make sure that context type authors didn't need to worry about the ids and registration system - and I think this has worked well to stop that implementation detail from leaking.

Could we do the same here please?

Reply:
Actually given that this is a connection level object and we have no arena considerations... we can probably move Destroy and Construct and name() to the traits type (delegating `name()` to the underlying type)... so:

```
template <typename T>
class ConnectionContextTraits {
 public:
   static uint16_t id() { return id_; }
   template <typename... Args>
   static T* Construct(Args&& args) { new T(std::forward<Args>(args)...); }
   static void Destruct(T* p) { delete p; }
   static const char* name(T* p) { return T::name(); }
};
```

and avoid users having to define these helper types at all.

Reply:
Done. I've updated to this format. I removed the name field though since it wasn't necessary.